tests/test_prompt_creation.py

```diff
@@ -41,172 +41,86 @@ def extract_unmasked_chunks(labels: List[int]) -> List[List[int]]:
-                EndToken.system.value,
-            ),
-            ({"role": "user", "content": "hello"}, EndToken.user.value),
-            (
-                {"role": "assistant", "content": "nice to meet you"},
-                EndToken.assistant.value,
-            ),
-            (
-                {
-                    "role": "assistant",
-                    "content": None,
-                    "function_call": {
-                        "name": "get_weather",
-                        "arguments": '{"city": "Hanoi"}',
-                    },
-                },
-                EndToken.function_call.value,
-            ),
-            (
-                {
-                    "role": "function",
-                    "content": '{"temperature": 30}',
-                    "name": "get_weather",
-                },
-                EndToken.function.value,
-            ),
-        ]
-
-        for message, stop_token in cases:
-            prompt = get_text_from_message(message).strip()
-            #  Check if prompt ends with stop_token
-            self.assertTrue(
-                prompt.endswith(stop_token),
-                f"`{prompt}` doesn't end with: `{stop_token}`",
-            )
-
-        # these cases don't require end_token at the end
-        edge_cases = [
-            ({"role": "user", "content": None}, EndToken.user.value),
-            ({"role": "assistant", "content": None}, EndToken.assistant.value),
-        ]
-
-        for message, stop_token in edge_cases:
-            prompt = get_text_from_message(message).strip()
-            #  Check if prompt doesn't endswith stop_token
-            self.assertFalse(prompt.endswith(stop_token), f"`{prompt}` ends with: `{stop_token}`")
-
     def test_prepare_training_inputs_fast_tokenizer(self):
         print("start testing fast tokenizer")
-        self.run_prepare_training_inputs(use_fast=True)
-    
+        self.run_prepare_training_inputs(
+            use_fast=True, pretrained="mistralai/Mistral-7B-v0.1"
+        )
+
     def test_prepare_training_inputs_normal_tokenizer(self):
         print("start testing normal tokenizer")
-        self.run_prepare_training_inputs(use_fast=False)
+        self.run_prepare_training_inputs(
+            use_fast=False, pretrained="mistralai/Mistral-7B-v0.1"
+        )
 
-    def run_prepare_training_inputs(self, use_fast: bool):
+    def run_prepare_training_inputs(self, use_fast: bool, pretrained: str):
         """this function is used to test function: prepare_training_inputs"""
         # note that must set legacy=True, read more: https://github.com/huggingface/transformers/issues/25176
         tokenizer_class = LlamaTokenizer
         if use_fast:
             tokenizer_class = LlamaTokenizerFast
-        tokenizer = tokenizer_class.from_pretrained("musabgultekin/functionary-7b-v1", legacy=True)
+        tokenizer = tokenizer_class.from_pretrained(pretrained, legacy=True)
+        tokenizer.pad_token = tokenizer.eos_token
```

Comment:
Actually each prompt template has **a function returning list of stop tokens** (this function https://github.com/MeetKai/functionary/blob/new_prompt_design/functionary/prompt.py#L27). The reason for a **list of stop tokens** instead of **a single token** because, previously, we have 2 stop tokens: <|END_OF_ASSISTANT|> or <|END_OF_FUNCTION_CALL|>. So if we want to set the eos_token in tokenizer, we don't know which to choose from the list of stop tokens.

Reply:
setting pad because mistal has no pad_token, so we sed pad_token=eos_token. In the training we already set this